src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,791 @@
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return list;
+}
+
+int QgsHanaConnection::getLayerSRID( const QgsHanaLayerProperty &layerProperty )
+{
+  if ( layerProperty.geometryColName.isEmpty() )
+    return -1;
+
+  int srid = -1;
+
+  try
+  {
+    PreparedStatementRef stmt;
+    if ( !layerProperty.isView )
+    {
+      const char *sql = "SELECT SRS_ID FROM SYS.ST_GEOMETRY_COLUMNS "
+                        "WHERE SCHEMA_NAME = ? AND TABLE_NAME = ? AND COLUMN_NAME = ?";
+      stmt = mConnection->prepareStatement( sql );
+      stmt->setNString( 1, NString( layerProperty.schemaName.toStdU16String() ) );
+      stmt->setNString( 2, NString( layerProperty.tableName.toStdU16String() ) );
+      stmt->setNString( 3, NString( layerProperty.geometryColName.toStdU16String() ) );
+    }
+    else
+    {
+      QString sql = QStringLiteral( "SELECT %1.ST_SRID() FROM %2.%3 WHERE %1 IS NOT NULL LIMIT %4" )
+                    .arg( QgsHanaUtils::quotedIdentifier( layerProperty.geometryColName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.schemaName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.tableName ),
+                          QString::number( GEOM_TYPE_SELECT_LIMIT ) );
+      stmt = mConnection->prepareStatement( QgsHanaUtils::toUtf16( sql ) );
+    }
+
+    int prevSrid = -1;
+    ResultSetRef rsSrid = stmt->executeQuery( );
+    while ( rsSrid->next() )
+    {
+      srid = *rsSrid->getInt( 1 );
+      if ( prevSrid != -1 && srid != prevSrid )
+      {
+        srid = -1;
+        break;
+      }
+      prevSrid = srid;
+    }
+
+    rsSrid->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return srid;
+}
+
+QStringList QgsHanaConnection::getLayerPrimaryeKeys( const QgsHanaLayerProperty &layerProperty )
+{
+  QStringList ret;
+
+  try
+  {
+    DatabaseMetaDataUnicodeRef dbmd = mConnection->getDatabaseMetaDataUnicode();
+    ResultSetRef rsPrimaryKeys = dbmd->getPrimaryKeys( nullptr,
+                                 QgsHanaUtils::toUtf16( layerProperty.schemaName ),
+                                 QgsHanaUtils::toUtf16( layerProperty.tableName ) );
+
+    while ( rsPrimaryKeys->next() )
+    {
+      auto keyName = rsPrimaryKeys->getNString( 4 );
+      QString clmName = QgsHanaUtils::toQString( keyName );
+      ResultSetRef rsColumns = dbmd->getColumns( nullptr,
+                               QgsHanaUtils::toUtf16( layerProperty.schemaName ),
+                               QgsHanaUtils::toUtf16( layerProperty.tableName ),
+                               QgsHanaUtils::toUtf16( clmName ) );
+
+      if ( rsColumns->next() )
+      {
+        Short dataType = rsColumns->getShort( 5 );
+        short dt = *dataType;
+        if ( dt == SQLDataTypes::TinyInt || dt == SQLDataTypes::SmallInt ||
+             dt == SQLDataTypes::Integer || dt == SQLDataTypes::BigInt )
+          ret << clmName;
```

Comment:
But what about tables that have a non-numeric primary key column? And what about tables that have a numeric + a non-numeric primary key column? In that case we would tell QGIS that the only primary key component is the numeric column. And then the database could return more than one row per partial primary key.

Reply:
Reworked in 2267454.